### ISSUE TITLE
Support for Keccak

### ISSUE THREAD

**Issue by kumavis** [multiformats/multihash#54]
We need support for base Keccak as well as the already added SHA-3 standard subset of Keccak.
Ethereum calls its hash algo of choice "SHA-3" but was chosen before the standard was settled, and is more accurately called "Keccak-256".

| Instance | Definition |
| --- | --- |
| eth_sha3(M) | Keccak[256](M, 256) |
| SHA3-224(M) | Keccak[448](M || 01, 224) |
| SHA3-256(M) | Keccak[512](M || 01, 256) |
| SHA3-384(M) | Keccak[768](M || 01, 384) |
| SHA3-512(M) | Keccak[1024](M || 01, 512) |
| SHAKE128(M, d) | Keccak[256](M || 1111, d) |
| SHAKE256(M, d) | Keccak[512](M || 1111, d) |

table from [sha-3 wiki](https://en.wikipedia.org/wiki/SHA-3)

further reading:
- http://ethereum.stackexchange.com/questions/550/which-cryptographic-hash-function-does-ethereum-use
- https://github.com/ethereum/EIPs/issues/59
- https://github.com/emn178/js-sha3
- http://emn178.github.io/online-tools/keccak_256.html
- http://emn178.github.io/online-tools/sha3_256.html

**Comment by kumavis** [comment]
relevant https://github.com/multiformats/multihash/issues/11

**Comment by kumavis** [comment]
cc @jbenet @diasdavid

**Comment by jbenet** [comment]
Yeah absolutely, we should add all these. @diasdavid is revamping multicodec table, maybe he can assign codes for them?

**Comment by diasdavid** [comment]
> Was just replying through IRC, but better to answer here

tl;dr; the multicodec table inside the multihash table is out of date

See https://github.com/multiformats/multicodec/tree/spec/update, it already has codes for the algos you are looking for. If I'm reading it correctly, there is a direct mapping in implementation between sha3 and Keccak, so, should we recognize both names, or will it be acceptable for devs to use the sha3 names, since their are standard by NIST?

With regards to eth_sha3, since it is a special variant, we can add it, can we call it `sha3-eth`, so that it is adequately namespaced like all the other sha3?

@jbenet can we get a review on https://github.com/multiformats/multicodec/pull/16, there has been some proposals with regards to multicodecs

**Comment by Kubuxu** [comment]
Its official name AFAIK, is keccak-256 so I would go with it.

**Comment by kumavis** [comment]
@diasdavid to be clear, `NIST standard SHA-3` is a *subset* of `keccak`. `eth-sha3` uses "base" `keccak_256`.
* you *can* perform `NIST standard SHA-3` if you have `keccak-*`.
* you *can* perform `eth-sha3` if you have `keccak-*`.
* you *can not* perform `eth-sha3` if you (only) have `NIST standard SHA-3`.

**Comment by kumavis** [comment]
Here is the PR for selecting a codec code https://github.com/multiformats/multicodec/pull/19

**Comment by axic** [comment]
NIST SHA3 and Keccak differ in the padding byte only, otherwise they are the same.

**Comment by axic** [comment]
@diasdavid:
> With regards to eth_sha3, since it is a special variant, we can add it, can we call it sha3-eth, so that it is adequately namespaced like all the other sha3?

It is not a special variant. Keccak is the hashing algorithm winning the SHA3 competition and for some reason the standard got the padding byte changed.

@Kubuxu is right, it should be called `keccak-<n>`

**Comment by diasdavid** [comment]
@axic understood, at first from a conversation, it sounded like eth_sha3 was not straight Keccak or sha3.

**Comment by Kubuxu** [comment]
I think yes.
